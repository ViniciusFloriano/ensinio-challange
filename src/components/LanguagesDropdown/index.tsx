import React from 'react';

import LanguagesDropdownItem, {
  LanguagesDropdownItemProps,
} from '../LanguagesDropdownItem';
import Container from './styles';

interface LanguagesDropdownProps {
  items: LanguagesDropdownItemProps[];
}

const LanguagesDropdown: React.FC<LanguagesDropdownProps> = ({ items }) => (
  <Container>
    {items && items.map((item) => <LanguagesDropdownItem {...item} />)}
  </Container>
);

export default LanguagesDropdown;
