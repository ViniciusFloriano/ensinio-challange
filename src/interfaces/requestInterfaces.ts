import { LanguagesProps } from './languageInterfaces';

export interface ItemProps {
  id: number;
  title: LanguagesProps;
  description: LanguagesProps;
}
