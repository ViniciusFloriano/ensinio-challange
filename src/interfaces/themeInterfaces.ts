export interface ThemeProps {
  colors: {
    white?: string;
    black?: string;
    gray?: string;
    lightGray?: string;
  };
  sizes: {
    xsmall: string | number;
    small: string | number;
    normal: string | number;
    large: string | number;
    xlarge: string | number;
    xxlarge: string | number;
  };
}
