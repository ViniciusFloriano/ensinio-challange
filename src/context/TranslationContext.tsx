import React, { createContext, useEffect, useState } from 'react';

import { LanguagesProps } from '../interfaces/languageInterfaces';
import { TranslationProps } from '../interfaces/requestInterfaces';
import { getTranslations } from '../services/requests';

type LanguagesType = keyof LanguagesProps;

interface TranslationContextProps {
  loading: boolean;
  error: boolean;
  currentLang: LanguagesType;
  setCurrentLang: (language: LanguagesType) => void;
  t: (title: string) => string;
  translationData: TranslationProps[];
}

const Context = createContext<TranslationContextProps>({
  loading: true,
  error: false,
  currentLang: 'pt',
  setCurrentLang: () => {},
  t: (title: string) => title,
  translationData: [],
});

function useTranslation() {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(true);
  const [translationData, setTranslationData] = useState<TranslationProps[]>(
    []
  );
  const [currentLang, setCurrentLangState] = useState<LanguagesType>('pt');

  useEffect(() => {
    getTranslations()
      .then((data) => {
        setTranslationData(data);
        setLoading(false);
      })
      .catch(() => {
        setError(true);
      });

    const languageStorage = localStorage.getItem('language');

    if (languageStorage) {
      setCurrentLangState(JSON.parse(languageStorage));
    } else {
      const navigatorLanguage = navigator.language;
      if (navigatorLanguage.includes('en')) {
        setCurrentLangState('en');
      } else if (navigatorLanguage.includes('pt')) {
        setCurrentLangState('pt');
      } else if (navigatorLanguage.includes('es')) {
        setCurrentLangState('es');
      }
    }
  }, []);

  function t(title: string) {
    const translatedText = translationData.find(
      (td) => td.title.toLowerCase() === title.toLowerCase()
    );

    if (!translatedText) {
      return 'Translation not find';
    }

    return translatedText.value[currentLang];
  }

  function setCurrentLang(language: LanguagesType) {
    localStorage.setItem('language', JSON.stringify(language));
    setCurrentLangState(language);
  }

  return { loading, error, currentLang, setCurrentLang, t, translationData };
}

const TranslationProvider: React.FC = ({ children }) => {
  const {
    loading,
    error,
    currentLang,
    setCurrentLang,
    t,
    translationData,
  } = useTranslation();

  return (
    <Context.Provider
      value={{
        loading,
        error,
        currentLang,
        setCurrentLang,
        t,
        translationData,
      }}
    >
      {children}
    </Context.Provider>
  );
};

export { Context as TranslationContext, TranslationProvider };
