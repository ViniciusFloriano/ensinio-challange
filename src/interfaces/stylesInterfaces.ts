import { ThemeProps } from './themeInterfaces';

export interface HeadingProps {
  color?: keyof ThemeProps['colors'];
  size?: keyof ThemeProps['sizes'];
  fontWeight?: 100 | 400 | 700;
  lineHeight?: string | number;
  level?: 1 | 2 | 3 | 4 | 5 | 6;
}
