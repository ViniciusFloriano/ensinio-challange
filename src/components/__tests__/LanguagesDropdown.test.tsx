import React from 'react';
import renderer from 'react-test-renderer';

import { cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { TranslationContext } from '../../context/TranslationContext';
import { LanguagesProps } from '../../interfaces/languageInterfaces';
import GlobalStyle from '../../styles/global';
import defaultTheme from '../../styles/themes';
import LanguagesDropdown from '../LanguagesDropdown';

it('renders correctly when there are no items', () => {
  const language: keyof LanguagesProps = 'pt';
  const translation = {
    loading: true,
    error: false,
    currentLang: language,
    setCurrentLang: (strLanguage: string) => strLanguage,
    t: (title: string) => title,
    translationData: [],
  };
  const languagesItems = [
    {
      icon: 'BrazilIcon',
      key: '01',
      title: 'PT',
      checked: translation.currentLang === 'pt',
      onClick: () => translation.setCurrentLang('pt'),
    },
    {
      icon: 'UsaIcon',
      key: '02',
      title: 'EN',
      checked: translation.currentLang !== 'pt',
      onClick: () => translation.setCurrentLang('en'),
    },
    {
      icon: 'SpainIcon',
      key: '03',
      title: 'ES',
      checked: translation.currentLang !== 'pt',
      onClick: () => translation.setCurrentLang('es'),
    },
  ];

  const tree = renderer
    .create(
      <TranslationContext.Provider value={translation}>
        <ThemeProvider theme={defaultTheme}>
          <LanguagesDropdown
            items={languagesItems}
          />
          <GlobalStyle />
        </ThemeProvider>
      </TranslationContext.Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
afterEach(cleanup);
