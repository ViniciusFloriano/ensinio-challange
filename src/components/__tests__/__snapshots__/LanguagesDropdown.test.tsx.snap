// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div
  className="sc-dlnjwi fEjFgp"
>
  <div
    aria-hidden="true"
    checked={true}
    className="sc-gtsrHT jFCtKs"
    onClick={[Function]}
  >
    <img
      alt="PT"
      src="BrazilIcon"
    />
    <p
      className="sc-bdnxRM gHVmtO"
      color="textTertiary"
      fontWeight={400}
      size="small"
    >
      PT
    </p>
    <img
      alt="PT"
      src="check.svg"
    />
  </div>
  <div
    aria-hidden="true"
    checked={false}
    className="sc-gtsrHT jDEFIP"
    onClick={[Function]}
  >
    <img
      alt="EN"
      src="UsaIcon"
    />
    <p
      className="sc-bdnxRM gHVmtO"
      color="textTertiary"
      fontWeight={400}
      size="small"
    >
      EN
    </p>
  </div>
  <div
    aria-hidden="true"
    checked={false}
    className="sc-gtsrHT jDEFIP"
    onClick={[Function]}
  >
    <img
      alt="ES"
      src="SpainIcon"
    />
    <p
      className="sc-bdnxRM gHVmtO"
      color="textTertiary"
      fontWeight={400}
      size="small"
    >
      ES
    </p>
  </div>
</div>
`;
